### ISSUE TITLE
Reveal known cards in opponent hand.

### ISSUE THREAD

**Issue by Epix37** [Epix37/Hearthstone-Deck-Tracker#2224]
Show cards created in the opponent hand in a similar fashion as cards that were created in the deck.

There is a chance this can be done by parsing the card text, but at least some of the cards will require hardcoding it.

Todo: make list of cards that create cards in hand.

**Comment by culinko** [comment]
Related: #961.

**Comment by HellbringerOnline** [comment]
It may be possible to use turn history?
![default](https://cloud.githubusercontent.com/assets/14059683/13818655/cdb95152-eba7-11e5-8674-b72f877aa1ab.PNG)

**Comment by Epix37** [comment]
That still wouldn't tell me what activating a certain card actually does. 

Using the turn history (and image recognition is not really a pretty thing) could only say *that* something happened, but not what. I can see from the logs that Antonidas was activated and that he created a card, but the logs do not say what that card is. Of course, we know because we understand the card text, HDT does not however. And while parsing the card text might work in some cases, there are others where there is no exact (or only an ambiguous) card specified in the text - spare part for example.

So basically, for every single card that does something like this I would have to say `if card X activates, reveal card Y`. Which is a lot of work to not only do initially, but maintain aswell, as new expansions will add cards that do similar things.

**Comment by jleclanche** [comment]
Looking over the fireplace implementations, this should be exhaustive (I didn't include adventure cards; also left out tavern brawl cards):

Specific cards:
* Play
 * Wild Growth (CS2_013): Excess Mana (CS2_013t)
* Triggers
 * Archmage Antonidas (EX1_559): Fireball (CS2_029)
 * Cutpurse (AT_031): The Coin (GAME_005)
 * Recruiter (AT_113): Squire (CS2_152)
 * Headcrack (EX1_137): Headcrack (EX1_137) - end of turn trigger, no source so hard to detect 
* Deathrattle
 * Tomb Pillager (LOE_012): The Coin (GAME_005)
 * Rhonin (AT_009): Arcane Missiles (EX1_277) (x3)
 * Explorer's Hat (LOE_105e - deathrattle from minion buff): Explorer's Hat (LOE_105)

Game state dependant:
* Play
 * Burgle (AT_033): Copy of card in player's deck  **or** GAME_005 (The Coin) if NEUTRAL
 * Nefarian (BRM_030): CardClass=CLASS(FRIENDLY_HERO) **or** BRM_030t (Tail Swipe) if NEUTRAL (x2)
 * Mind Vision (CS2_003): Copy of card in player's hand
 * Thoughtsteal (EX1_339): Copy of card in player's deck (x2)
 * Convert (AT_015): Copy of TARGET
 * Sense Demons (EX1_317): EX1_317t if created, CardType=MINION Race=DEMON if moved from DECK
* Triggers
 * Trade Prince Gallywix (GVG_028): Copy of played card
 * Duplicate (FP1_018 - secret): Copy of entity that died

Random collectibles:
* Play
 * Webspinner (FP1_011): CardType=MINION Race=BEAST
 * Unstable Portal (GVG_003): CardType=MINION
 * Neptulon (GVG_042): CardType=MINION Race=MURLOC (x4)
 * Grand Crusader (AT_118): CardClass=PALADIN
 * Nexus-Champion Saraad (AT_127): CardType=SPELL CardClass=CLASS(FRIENDLY_HERO)
 * Lock and Load (AT_061e - player buff): CardClass=HUNTER

Random misc:
* Triggers
 * Ysera (EX1_572): CardClass=DREAM
 * Mech-Bear-Cat (GVG_084): SPARE_PART
* Play
 * Tinkertown Technician (GVG_102): SPARE_PART
 * Toshley (GVG_115): SPARE_PART
* Deathrattle
 * Toshley (GVG_115): SPARE_PART
 * Clockwork Gnome (GVG_082): SPARE_PART

**Comment by jleclanche** [comment]
Not included in this list: Tracking, Captain's Parrot.

**Comment by jleclanche** [comment]
Also not included: All the discover effects. Easy enough to add those.

**Comment by jleclanche** [comment]
Also tracking: Golden Monkey (Morph all in deck, CardType=MINION Rarity=LEGENDARY). For your own cards, you can already get all the IDs I think.

**Comment by culinko** [comment]
Also Headcrack is the same ID that was moved to graveyard, not a new entity (same as Malorne, Entomb). Not sure about Explorer's Hat. More created cards also here: https://github.com/HearthSim/hs-bugs/issues/375.

**Comment by culinko** [comment]
* Play:
- [ ] Astral Communion: Excess Mana -> hand (if at 10 max? mana)
- [ ] Echo of Medivh: minions from board from oldest to newest -> hand
- [ ] Elise Starseeker: Map to the Golden Monkey -> deck
- [ ] Excavated Evil: Excavated Evil -> opponent's deck
- [ ] Forgotten Torch: Roaring Torch -> deck
- [ ] Gang Up: 3 copies of target minion -> deck
- [ ] King Mukla: 2 Bananas -> hand
- [ ] Map to the Golden Monkey: Golden Monkey -> deck (can draw it immediately)
- [ ] Mulch: minion -> hand
- [ ] Spellslinger: spell -> hand
- [ ] Elite Tauren Chieftain: Power Chord card -> hand

Trigger:
- [ ] Gazlowe: Mech -> hand if played card is a 1 mana spell (counts mana reduction too)
- [ ] Lorewalker Cho: played spell -> hand

Discover:
- [ ] Arch-Thief Rafaam: Artifact -> hand
- [ ] Dark Peddler: 1 cost card -> hand
- [ ] Ethereal Conjurer: spell -> hand
- [ ] Gorillabot A-3: Mech -> hand
- [ ] Jeweled Scarab: 3 cost card -> hand
- [ ] Museum Curator: deathrattle minion -> hand
- [ ] Raven Idol: card -> hand
- [ ] Tomb Spider: Beast -> hand

**Comment by HellbringerOnline** [comment]
Priest's Shadow Madness card adds a minion to the opponents deck without a gift icon

**Comment by HellbringerOnline** [comment]
Unstable Portal card adds a minion to the opponents deck without a gift icon (not sure). This minion shown in the opp. deck in replay.

**Comment by Epix37** [comment]
@HellbringerOnline this is not the right issue for this. Please make a new one and provide screenshots / logs.

**Comment by cachirro** [comment]
thanks for #2838, all good things to wait for
